File: src/main/java/com/hjj/apiserver/service/AccountBookService.java
package com.hjj.apiserver.service;

import com.hjj.apiserver.common.exception.UserNotFoundException;
import com.hjj.apiserver.domain.AccountBookEntityJava;
import com.hjj.apiserver.domain.PurchaseEntityJava;
import com.hjj.apiserver.dto.AccountBookDto;
import com.hjj.apiserver.dto.AccountBookUserDto;
import com.hjj.apiserver.repositroy.PurchaseRepository;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;


@Transactional(readOnly = true)
@AllArgsConstructor
public class AccountBookService {

    private final AccountBookUserRepository accountBookUserRepository;
    private final AccountBookRepository accountBookRepository;
    private final CategoryService categoryService;
    private final PurchaseRepository purchaseRepository;
    private final UserRepository userRepository;
    private final CardService cardService;


    @Transactional(readOnly = false, rollbackFor = Exception.class)
    public void addAccountBook(UserEntity user, AccountBookDto accountBookDto) throws UserNotFoundException {

        AccountBookEntityJava accountBookEntity = accountBookDto.toEntity();
        accountBookRepository.save(accountBookEntity);

        /* 가계부 생성후 유저와 매핑 해주어야 한다. (기본 권한은 USER이므로 OWNER를 넣는다. */
        AccountBookUserDto accountBookUserDto = new AccountBookUserDto();
        accountBookUserDto.setAccountBookEntity(accountBookEntity);
        accountBookUserDto.setAccountRole(AccountBookUserEntityJava.AccountRole.OWNER);
        accountBookUserDto.setUserEntity(user);
        accountBookUserDto.setBackGroundColor(accountBookDto.getBackGroundColor());
        accountBookUserDto.setColor(accountBookDto.getColor());
        AccountBookUserEntityJava accountBookUserEntity = accountBookUserDto.toEntity();
        accountBookUserRepository.save(accountBookUserEntity);

        categoryService.addBasicCategory(accountBookEntity);
    }

    public AccountBookDto.ResponseAccountBookDetail findAccountBookDetail(AccountBookDto accountBookDto) throws Exception {
        AccountBookDto.ResponseAccountBookDetail responseAccountBookDetail = new AccountBookDto.ResponseAccountBookDetail();
        responseAccountBookDetail.setAccountBookName(accountBookRepository.getById(accountBookDto.getAccountBookNo()).getAccountBookName());
        responseAccountBookDetail.setCardList(cardService.selectCardList(accountBookDto.getUserNo()));
        responseAccountBookDetail.setCategoryList(categoryService.findAllCategory(accountBookDto.getUserNo(), accountBookDto.getAccountBookNo()));

        return responseAccountBookDetail;
    }


    public List<AccountBookDto.ResponseAccountBookFindAll> findAllAccountBook(AccountBookDto accountBookDto) {
        /* userNo으로 가계부 <-> 유저 매핑 테이블 조회 */
        List<AccountBookUserEntityJava> accountBookUserEntityList =  accountBookUserRepository.findEntityGraphByUserEntity_userNo(accountBookDto.getUserNo());
        List<Long> accountBookNoList = new ArrayList<>();
        List<AccountBookDto.ResponseAccountBookFindAll> responseAccountBookFindAllList = new ArrayList<>();

        accountBookUserEntityList.stream().forEach(accountBookUserEntity -> accountBookNoList.add(accountBookUserEntity.getAccountBookEntity().getAccountBookNo()));

        List<PurchaseEntityJava> purchaseEntityList = purchaseRepository.findAllEntityGraphByPurchaseDateBetweenAndUserEntity_UserNoOrderByPurchaseDateDesc(accountBookDto.getStartDate(), accountBookDto.getEndDate(), accountBookDto.getUserNo());
        List<AccountBookUserEntityJava> tempAccountBookUserList = accountBookUserRepository.findEntityGraphByAccountBookEntity_accountBookNoIn(accountBookNoList);


        accountBookUserEntityList.stream().forEach(accountBookUserEntity -> {
            Long accountBookNo = accountBookUserEntity.getAccountBookEntity().getAccountBookNo();
            AccountBookDto.ResponseAccountBookFindAll responseAccountBookFindAll = new AccountBookDto.ResponseAccountBookFindAll();
            responseAccountBookFindAll.setAccountBookNo(accountBookNo);
            responseAccountBookFindAll.setAccountBookName(accountBookUserEntity.getAccountBookEntity().getAccountBookName());
            responseAccountBookFindAll.setAccountBookDesc(accountBookUserEntity.getAccountBookEntity().getAccountBookDesc());
            responseAccountBookFindAll.setAccountRole(accountBookUserEntity.getAccountRole());
            responseAccountBookFindAll.setBackGroundColor(accountBookUserEntity.getBackGroundColor());
            responseAccountBookFindAll.setColor(accountBookUserEntity.getColor());

            int totalIncomeAmount = 0;
            int totalOutgoingAmount = 0;
            for (PurchaseEntityJava purchaseEntity : purchaseEntityList) {
                if(purchaseEntity.getPurchaseType().equals(PurchaseEntityJava.PurchaseType.INCOME) && purchaseEntity.getAccountBookEntity().getAccountBookNo() == accountBookNo){
                    totalIncomeAmount += purchaseEntity.getPrice();
                }else if(purchaseEntity.getPurchaseType().equals(PurchaseEntityJava.PurchaseType.OUTGOING) && purchaseEntity.getAccountBookEntity().getAccountBookNo() == accountBookNo){
                    totalOutgoingAmount += purchaseEntity.getPrice();
                }

            }
            responseAccountBookFindAll.setTotalIncomeAmount(totalIncomeAmount);
            responseAccountBookFindAll.setTotalOutgoingAmount(totalOutgoingAmount);

            List<AccountBookDto.ResponseAccountBookFindAll.JoinedUser> joinedUserList = new ArrayList<>();

            tempAccountBookUserList.stream().forEach(tempAccountBookEntity ->{
                if(accountBookNo == tempAccountBookEntity.getAccountBookEntity().getAccountBookNo()){
                    AccountBookDto.ResponseAccountBookFindAll.JoinedUser joinedUser = new AccountBookDto.ResponseAccountBookFindAll.JoinedUser();
                    joinedUser.setUserNo(tempAccountBookEntity.getUserEntity().getUserNo());
                    joinedUser.setNickName(tempAccountBookEntity.getUserEntity().getNickName());
                    joinedUser.setPicture(tempAccountBookEntity.getUserEntity().getPicture());
                    joinedUserList.add(joinedUser);
                }
            });

            responseAccountBookFindAll.setJoinedUserList(joinedUserList);
            responseAccountBookFindAllList.add(responseAccountBookFindAll);

        });

        return responseAccountBookFindAllList;
    }
}

File: src/main/kotlin/com/hjj/apiserver/service/AccountBookService.kt
package com.hjj.apiserver.service

import com.hjj.apiserver.domain.accountbook.AccountBook
import com.hjj.apiserver.domain.accountbook.AccountBookUser
import com.hjj.apiserver.domain.accountbook.AccountRole
import com.hjj.apiserver.domain.category.Category
import com.hjj.apiserver.dto.accountbook.request.AccountBookAddRequest
import com.hjj.apiserver.dto.accountbook.response.AccountBookDetailResponse
import com.hjj.apiserver.dto.accountbook.response.AccountBookFindAllResponse
import com.hjj.apiserver.repository.accountbook.AccountBookRepository
import com.hjj.apiserver.repository.accountbook.AccountBookUserRepository
import com.hjj.apiserver.repository.card.CardRepository
import com.hjj.apiserver.repository.user.UserRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class AccountBookService(
    private val accountBookUserRepository: AccountBookUserRepository,
    private val accountBookRepository: AccountBookRepository,
    private val categoryService: CategoryService,
    private val cardRepository: CardRepository,
    private val userRepository: UserRepository,
) {

    @Transactional(readOnly = false, rollbackFor = [Exception::class])
    fun addAccountBook(userNo: Long, request: AccountBookAddRequest): AccountBook {

        val newAccountBook = AccountBook(
            accountBookName = request.accountBookName,
            accountBookDesc = request.accountBookDesc,
        )

        val savedAccountBook = accountBookRepository.save(newAccountBook)
        accountBookUserRepository.save(
            AccountBookUser(
                accountBook = newAccountBook,
                user = userRepository.getById(userNo),
                accountRole = AccountRole.OWNER,
                backGroundColor = request.backGroundColor,
                color = request.color,
            )
        )
        categoryService.addBasicCategory(newAccountBook)
        return savedAccountBook
    }

    fun findAccountBookDetail(accountBookNo: Long, userNo: Long): AccountBookDetailResponse {
        val accountBook =
            accountBookRepository.findAccountBook(userNo, accountBookNo, listOf(AccountRole.MEMBER, AccountRole.OWNER))
                ?: throw IllegalArgumentException()

        val categories: MutableList<AccountBookDetailResponse.CategoryDetail> = mutableListOf()

        for (category in accountBook.categories) {
            if (category.parentCategory == null) {
                categories.add(AccountBookDetailResponse.CategoryDetail(
                    categoryNo = category.categoryNo,
                    categoryName = category.categoryName,
                    categoryIcon = category.categoryIcon,
                    childCategories = category.childCategories.map { childCategory: Category ->
                        AccountBookDetailResponse.ChildrenCategory(
                            categoryNo = childCategory.categoryNo,
                            categoryName = childCategory.categoryName,
                            categoryIcon = childCategory.categoryIcon,
                            parentCategoryNo = childCategory.parentCategory!!.categoryNo
                        )
                    }
                )
                )
            }
        }


        return AccountBookDetailResponse(
            accountBookNo = accountBook.accountBookNo!!,
            accountBookName = accountBook.accountBookName,
            accountBookDesc = accountBook.accountBookDesc,
            accountRole = accountBook.accountBookUserList.find { it.user.userNo == userNo }!!.accountRole,
            createdDate = accountBook.createdDate,
            cards = cardRepository.findByUser_UserNoAndDeleteYn(userNo).map {
                AccountBookDetailResponse.CardDetail(
                    cardNo = it.cardNo!!,
                    cardName = it.cardName
                )
            },
            categories = categories
        )
    }

    fun findAllAccountBook(userNo: Long): List<AccountBookFindAllResponse> {
        return accountBookUserRepository.findAllAccountBookByUserNo(userNo)
    }
}